### PR TITLE
Glob 50543/treat complex comprehensions

### PR DESCRIPTION
**Pull request by marcglobality**
Finally treat most of the complex comprehensions + added some refresher of black in the README (GLOB-50678)

## Why?

Because we still had some uncovered comprehensions (now hopefully most typical cases are covered)

## What?

See changes in comprehensions_input/output for all new cases covered + the ones not covered in known_failed_comprehensions

## Checklist

* [ ] Functionality has been tested end-to-end in a local environment.
* [ ] All changes are covered by automated tests including edge cases.
* [ ] API is backwards-compatible for upstream and downstream services.
* [ ] Relevant documentation (including comments) has been added or updated.
* [ ] I'm standing by to check the fix after deployment.

## References

* Link to PRs in other services
* Link to Sentry error
* Link to related documentation
* [Workflows and git flows](https://globality.atlassian.net/wiki/spaces/DEV/pages/1109493223/Workflows+and+git+flows)
* [How to do a great code review](https://globality.atlassian.net/wiki/spaces/DEV/pages/933758735/How+to+do+a+great+code+review)
* [[PSA] - Marquez Dev Builds](https://globality.atlassian.net/wiki/spaces/D/pages/937004290/PSA+-+Marquez+Dev+Builds)
* [Database Naming Convention](https://globality.atlassian.net/wiki/spaces/AP/pages/996442685/Database+Naming+Convention+and+Design+Guideline)
* [Schema Migrations](https://globality.atlassian.net/wiki/spaces/D/pages/1020761227/Schema+Migrations)
* Link to other reference

### CODE REVIEW

File: globality_black/comprehensions.py [hunk 1]
```diff
@@ -22,33 +22,52 @@ def reformat_comprehension(comp_for: PythonNode):
...
+    We explode if at least one of the following requirements are met:
+        - dict comprehension
+        - there is an if
+        - there is a ternary operator
+        - there are multiple for's
+
+    We don't explode:
+        - comprehension inside of comprehension (the inner one will be kept as is)
+        - if multiple for's, the first one already explodes the rest
+        (with `set_prefix_for_all_last_children`), so they are ignored
```
Comment (tomaszpietruszka-globality) [inline globality_black/comprehensions.py:34]: Not sure I understand

File: globality_black/comprehensions.py [hunk 2]
```diff
@@ -22,33 +22,52 @@ def reformat_comprehension(comp_for: PythonNode):
...
 
-    is_dict = (
-        comp_for.parent.type == ParsoTypes.DICTORSETMAKER.value
-        and comp_for.parent.children[1] == ":"
-    )
+    comp = comp_for.parent
+
+    is_dict = comp.type == ParsoTypes.DICTORSETMAKER.value and comp.children[1] == ":"
     ends_with_if = comp_for.children[-1].type == ParsoTypes.COMP_IF.value
     ends_with_for = comp_for.children[-1].type == ParsoTypes.SYNC_COMP_FOR.value
```
Comment (tomaszpietruszka-globality) [inline globality_black/comprehensions.py:43]: Don't all comprehensions end with a ` for x in y`, except those that add an `if` afterwards? not sure I get it

File: globality_black/comprehensions.py [hunk 3]
```diff
@@ -83,18 +96,17 @@ def set_prefix(element: PythonNode, prefix: str):
         leaf.prefix = leaf.prefix[:last_eol_position] + prefix
 
 
-def set_prefix_for_all_comp_if(comp_if: PythonNode, prefix: str):
+def set_prefix_for_all_last_children(comp_for: PythonNode, prefix: str):
     """
-    indent if in comprehension + all other if's
-    unfortunately parso treates each new comp_if after the first one (if any) as a child of the
+    indent for in comprehension + all for and if underneath
+    unfortunately parso treates each new comp_for and comp_if after the comp_for (if any) as a child
```
Comment (tomaszpietruszka-globality) [inline globality_black/comprehensions.py:102]:
```suggestion
    unfortunately parso treats each new comp_for and comp_if after the comp_for (if any) as a child
```

**Review by tomaszpietruszka-globality — APPROVED** [review]
LGTM with minor requests for clarification

File: globality_black/comprehensions.py (hunk 2 above)
Comment (marcglobality) [reply globality_black/comprehensions.py:43]: so `comp` is the whole comp, `comp_for` is e.g. `for x in y`. Comp-for's with multiple for's end (the last children is) a for, but not the simple ones

File: globality_black/comprehensions.py (hunk 3 above)
Comment (marcglobality) [reply globality_black/comprehensions.py:102]: 😱 👍

File: globality_black/comprehensions.py (hunk 1 above)
Comment (marcglobality) [reply globality_black/comprehensions.py:35]: explained offline

File: globality_black/tests/fixtures/known_failed_comprehensions_input.txt [hunk 4]
```diff
@@ -1,33 +1,6 @@
-x = [i for i in range(6) for j in range(i + 3)]
-
-x = [i for i in range(59) for j in range(i + 1) if i < j]
-
-x = [[i for i in range(7)] for j in range(10)]
-
+# this one is partially covered. We explode every comprehension having as value a comprehension,
```
Comment (ml-globality) [inline globality_black/tests/fixtures/known_failed_comprehensions_input.txt:1]: Is this a list of comprehensions that globality-black still doesn't support? Good to see it decrease! Maybe changed the first sentence to say `We only have partial coverage of nested comprehensions` to be more clear.

**Review by ml-globality — APPROVED** [review]
LGTM with small question

File: globality_black/tests/fixtures/known_failed_comprehensions_input.txt (hunk 4 above)
Comment (marcglobality) [reply globality_black/tests/fixtures/known_failed_comprehensions_input.txt]: as talked offline, updated
